BUILD/CONFIG: added module check script

hoopoepg opened the pull request:
- added helper script UCX_CHECK_MODULE to automatically detect modules
  installed on system

Commit:
BUILD/CONFIG: added module check script

- added helper script UCX_CHECK_MODULE to automatically detect modules
  installed on system

rakhmets commented on config/m4/module.m4:
```diff
@@ -0,0 +1,437 @@
+             AS_IF([test "x$2" != "x"], [CPPFLAGS="$2 $CPPFLAGS"])
+
+             m4_foreach([header], [$1],
+                 [ dnl have to clean check header cache to allov multiple checks
```
Typo 'allow'. The same issue at lines 164, 196.

hoopoepg replied:
yep, typo. fixed

rakhmets commented on config/m4/module.m4:
```diff
@@ -0,0 +1,437 @@
+#
+# Copyright (C) 2022 Nvidia Corporation. All Rights Reserved.
```
Copyright (C) 2022 NVIDIA CORPORATION & AFFILIATES. ALL RIGHTS RESERVED.

hoopoepg replied:
fixed

Commit:
BUILD/CONFIG: added module check script

- fixed typos

rakhmets commented on config/m4/module.m4:
```diff
@@ -0,0 +1,437 @@
+         ])
+
+# Check if module available
+# Result: set variables prefix_CPPFLAGS and prefix_LIBS
```
We use prefix_LIBS and prefix_LDFLAGS when configuring some modules.
E.g. prefix_LIBS contains list of libraries: -lmylib, prefix_LDFLAGS contains path to library: -L$path_to_my_lib.
Perhaps this is a better approach. What do you think about it?

hoopoepg replied:
is there any case when LIBS is used without LDFLAGS?

hoopoepg replied:
I used same as for PKG_CHECK_MODULES variables to store results

rakhmets replied:
> is there any case when LIBS is used without LDFLAGS?

Actually, prefix_LDFLAGS usually is added to module_la_LDFLAGS/bin_LDFLAGS, prefix_LIBS usually is added to module_la_LIBADD/bin_LIBADD in .am files.

hoopoepg replied:
the problem is in using PKG_CHECK_MODULES macro to identify module: it uses LIBS/CFLAGS to store results. I used same way to unify results variables

rakhmets replied:
I see...

(... the macro will set up two variables ... _CFLAGS and _LIBS. The names of these variables can be somewhat misleading, since the former will generally provide the flags to pass to the preprocessor ... such as include paths and macro definitions, and the latter will provide the library paths as well as the libraries themselves ...)

Therefore, it won't be trivial to rewrite existing configs.

Commit:
BUILD/CONFIG: added module check script

- unify result variables with PKG_CHECK_MODULES

Commit:
BUILD/CONFIG: added module check script

- added alternative module directories
- XPMEM detection is implemented using new module search engine

hoopoepg commented:
@yosefe ok to merge innovations?

yosefe commented on config/m4/module.m4:
```diff
@@ -0,0 +1,484 @@
+# helper macro
+m4_define([_CHECK_MODULE_MAKE_HELP_STRING], [m4_argn($1, $2)$3])
+    
+m4_ifndef([PKG_CHECK_MODULES_STATIC],
```
why need m4_ifndef?

yosefe replied:
pls add this explanation as a comment

yosefe commented on config/m4/module.m4:
```diff
@@ -0,0 +1,484 @@
+          ])
+
+#
+# Init progress
```
documentation and name are not clear.
what is this macro doing?

yosefe commented on config/m4/module.m4:
```diff
@@ -0,0 +1,484 @@
+m4_define([_CHECK_MODULE_RUN],
+          [
+             mod_run_happy=no
+
```
remove space line

yosefe commented on config/m4/module.m4:
```diff
@@ -0,0 +1,484 @@
+          ])
+
+#
+# Run test if status is in allowed list
+#
+# Usage: _CHECK_MODULE_RUN([status], [allowed-statuses], [macro], [if-no-macro])
```
name and doc not clear..
did to mean to check if $1 is equal to any of the values listed in $2 ?

yosefe commented on config/m4/module.m4:
```diff
@@ -0,0 +1,484 @@
+
+             m4_foreach([stat], [$2],
+                 [
+                     AS_IF([test "$1" = "stat"], [mod_run_happy=yes])dnl
```
why need dnl?

yosefe replied:
so it's needed for commands inside m4_foreach?
maybe use something like m4_set_contains instead of iteration?

yosefe replied:
BTW maybe writing the whole m4_foreach in a single line would help and be easier to read

yosefe commented on config/m4/module.m4:
```diff
@@ -0,0 +1,484 @@
+          ])
+
+#
+# Check if all directories are exist
```
Check if all directories listed in $1 exist

yosefe commented on config/m4/module.m4:
```diff
@@ -0,0 +1,484 @@
+             AS_IF([test "x$1" != "x"], [module_check_dir_happy=yes], [module_check_dir_happy=no])dnl
+             m4_foreach([dir], [$1],
+                        [
+                            AS_IF([test -d "dir"], [], [module_check_dir_happy=no])dnl
```
why nee dnl?

yosefe commented on config/m4/module.m4:
```diff
@@ -0,0 +1,484 @@
+         ])
+
+#
+# Fixup for lib path
+#
+# Usage: _CHECK_MODULE_FIXUP_LIB([lib], [dest_var], [if-success], [if-not])
```
documentation and name partial and unclear

yosefe commented on config/m4/module.m4:
```diff
@@ -0,0 +1,484 @@
+                                 [$2=yes],   # yes
+                                 [$2=skip],  # no
+                                 [$2=guess], # guess
+                                 [$2=skip],  # empty
```
it means empty is same as no?
is it the current behavior? i think today empty behaves as guess

yosefe replied:
why would this happen, since we define AC_ARG_WITH in our macro?
maybe we could remove this, for simplicity?

yosefe commented on config/m4/module.m4:
```diff
@@ -0,0 +1,484 @@
+#
+# Usage: _CHECK_MODULE_IS_YNG([value], [if-yes], [if-no], [if-guess], [if-empty], [if-else])
+#
+AC_DEFUN([_CHECK_MODULE_IS_YNG],
```
maybe unite this with _CHECK_MODULE_INIT_PROGRESS ?

yosefe commented on config/m4/module.m4:
```diff
@@ -0,0 +1,484 @@
+#
+# Check module header
+#
+# Usage: _CHECK_MODULE_HEADER([header], [cppflags], [if-found], [if-not-found], [includes])
```
can there be several headers?
it should be _CHECK_MODULE_HEADERS([headers] ...

yosefe commented on config/m4/module.m4:
```diff
@@ -0,0 +1,484 @@
+                  unset header_cache
+                  mod_header_name=header
+                  AC_CHECK_HEADER([$mod_header_name], [], [mod_header_happy=no], [$5])
+                  AS_VAR_POPDEF([header_cache])])dnl
```
why need dnl?

yosefe commented on config/m4/module.m4:
```diff
@@ -0,0 +1,484 @@
+             CPPFLAGS=$mod_decls_saved_CPPFLAGS
+
+             # in case if decls is empty list - then assume that test is successfull
+             AS_IF([test $mod_decls_happy = yes -o "x$1" = "x"], [$3], [$4])
```
maybe place the AC_CHECK_DECLS above under AS_IF([test -n "$1"])

yosefe commented on config/m4/module.m4:
```diff
@@ -0,0 +1,484 @@
+             mod_decls_saved_CPPFLAGS=$CPPFLAGS
+             mod_decls_happy=yes
+
+             AS_IF([test "x$2" != "x"], [CPPFLAGS="$2 $CPPFLAGS"])
```
why need AS_IF?

yosefe commented on config/m4/module.m4:
```diff
@@ -0,0 +1,484 @@
+             mod_header_saved_CPPFLAGS=$CPPFLAGS
+             mod_header_happy=yes
+
+             AS_IF([test "x$2" != "x"], [CPPFLAGS="$2 $CPPFLAGS"])
```
why need AS_IF?

yosefe commented on config/m4/module.m4:
```diff
@@ -0,0 +1,484 @@
+             mod_libs_saved_LIBS=$LIBS # save it because AC_SEARCH_LIBS breaks LIBS var
+             mod_libs_happy=yes
+
+             AS_IF([test "x$2" != "x"], [LDFLAGS="$2 $LDFLAGS"])
```
why need AS_IF?

yosefe commented on config/m4/module.m4:
```diff
@@ -0,0 +1,484 @@
+AC_DEFUN([_CHECK_MODULE_API],
+         [
+             _CHECK_MODULE_DEFVAR([$5])
+
```
IMO better remove such spacelines (in multiple places)

yosefe commented on config/m4/module.m4:
```diff
@@ -0,0 +1,484 @@
+             mod_pkg_saved_PKG_CONFIG_PATH="$PKG_CONFIG_PATH"
+             mod_package_happy=yes
+
+             AS_IF([test "x$1" = "x"], [mod_package_happy=no])
```
maybe use "test -n" and "test -z" to test for empty strings

yosefe commented on config/m4/module.m4:
```diff
@@ -0,0 +1,484 @@
+                   [
+                       _CHECK_MODULE_IS_DIR([$3/pkgconfig],
+                                            [export PKG_CONFIG_PATH="$3/pkgconfig:$PKG_CONFIG_PATH"],
+                                            [mod_package_happy=no])
```
why must require the module has $libdir/pkgconfig?

yosefe replied:
some 3rd party libs may provide only .h and .so/.a but not pkgconfig

yosefe commented on config/m4/module.m4:
```diff
@@ -0,0 +1,484 @@
+# Usage: _CHECK_MODULE_PACKAGE([pkgid], [prefix], [libdir], [static], [if-found], [if-not-found],
+#                              [[header], [decls], [includes], [libs], [funcs], [ldflags]])
+#
+# DO NOT!!!!! use macro substitution for 'prefix' var of  _CHECK_MODULE_PACKAGE
```
maybe use AS_IF instead of PKG_CHECK_MODULES$4

yosefe replied:
IMO it makes harder to track the macros this way

yosefe commented on config/m4/module.m4:
```diff
@@ -0,0 +1,484 @@
+             export PKG_CONFIG_PATH="$mod_pkg_saved_PKG_CONFIG_PATH"
+
+             AS_IF([test $mod_package_happy = yes],
+                   [
+                       _CHECK_MODULE_API([$[$2_CFLAGS]], [$[$2_LIBS]],
```
compact lines

yosefe replied:
```m4
             AS_IF([test $mod_package_happy = yes],
                   [_CHECK_MODULE_API([$[$2_CFLAGS]], [$[$2_LIBS]], [],
                                      [mod_package_happy=no], [$7])])
```

yosefe commented on config/m4/module.m4:
```diff
@@ -0,0 +1,484 @@
+#       1. check for lib using declarations/link
+#       2. check for package
+#
+# Usage: UCX_CHECK_MODULE([id], [name], [prefix], [pkgname],
+#                         [[doc], [doc-lib], [doc-static or '-' to skip], [doc-failed]],
+#                         [action-if-found], [action-if-not-found],
+#                         [[headers], [decls], [includes], [libs - space separated], [funcs], [otherlibs]],
+#                         [alt-dirs])
```
pls document the parameters

yosefe commented on config/m4/module.m4:
```diff
@@ -0,0 +1,484 @@
+#                         [alt-dirs])
+#
+# Example:
+# m4_define([t1_arg], [[ucp/api/ucp.h, ucs/config/global_opts.h],
```
why need t1_arg and not pass all as parameters?

yosefe commented on config/m4/module.m4:
```diff
@@ -0,0 +1,484 @@
+
+             mod_progress=start
+
+             _CHECK_MODULE_RUN([$mod_progress], [start],
```
i think there are too many steps in this macro, why need so many?
maybe add another helper macro like checking static lib?

yosefe replied:
seems too many... i would expect something like:
1. check headers
2. check static libs
3. check dynamic libs

and for each such step: 1. search user specified dir, 2. try pkgconfig, 3. use default paths

Maybe we can express this structure with a more generic macro internal helper so it will be more "compact" code?

hoopoepg replied:
every test has own structure: static lib is checked by pkgconfig first because it automatically detects for dependencies and if it failed - try to check library directly. dynamic library is detected by library first to provide compatibility with "legacy" way of module check and if failed - try to use pkgconfig. if user specify any directory then test must fail in case if directory inaccessible. if user didn't set any directory and module could not be found in default location then alternative directories should be checked and failures on such checks should be ignored. in case if module is detected using PKG_CHECK_MODULES macro then _CFLAGS and _LIBS variables should not be updated manually because it is updated by script. there are tons of combinations of tests and user input and adding any structure here just adds very nested conditions.

yosefe commented on configure.ac:
```diff
@@ -88,6 +88,7 @@ AC_FUNC_STRERROR_R
 
 AC_PATH_TOOL([PKG_CONFIG], [pkg-config], [pkg-config])
 
+PKG_PROG_PKG_CONFIG
```
do we still need line 89?

yosefe commented on src/uct/sm/mm/xpmem/configure.m4:
```diff
@@ -5,53 +5,17 @@
-       CPPFLAGS="$save_CPPFLAGS"
-      ]
-     )
+UCX_CHECK_MODULE([xpmem], [XPMEM], [XPMEM], [xpmem],
+                 [[Enable the use of XPMEM], [Search XPMEM library in DIR], [-],
+                  [XPMEM could not be found]],
+                  [xpmem_happy=yes], [],
+                  [[xpmem.h], [xpmem_version], [[#include <xpmem.h>]], [xpmem], [xpmem_init], []],
+                  [/opt/xpmem, /usr/local, /usr/local/xpmem])
```
maybe some parameters here can be deduced from others and removed:
- "XPMEM could not be found" -> "$2 could not be found"
- "xpmem_happy=yes" -> "$1_happy=yes"
- "Enable the use of XPMEM" -> "Enable the use of $2"
- "Search XPMEM library in DIR" -> "Search $2 library in DIR"

yosefe commented on src/uct/sm/mm/xpmem/configure.m4:
```diff
@@ -5,53 +5,17 @@
 
 AS_IF([test "x$xpmem_happy" = "xyes"], [uct_modules="${uct_modules}:xpmem"])
 AM_CONDITIONAL([HAVE_XPMEM], [test "x$xpmem_happy" != "xno"])
+AC_SUBST(XPMEM_CFLAGS)
+AC_SUBST(XPMEM_LIBS)
```
can it be part of the UCX_CHECK_MODULE macro?

yosefe replied:
IMO it's OK as long as documented properly

hoopoepg replied:
> IMO it's OK as long as documented properly

ok

Commit:
BUILD/CONFIG: added module check script

- refactoring of scripts

hoopoepg commented:
@yosefe ok to squash?

yosefe commented on config/m4/module.m4:
```diff
@@ -36,25 +37,32 @@ m4_define([_CHECK_MODULE_UNDEFVAR],
 #
-# Configure module directories
+# Configure module directories:
+#     inc=$mod-dir/include
+#     incflag=-I$mod-dir/include
+#     lib=$mod-dir/lib64 or $mod-dir/lib
+#     libflag=-L$lib (from previous step)
+#     if any of directories is missing - call if-failed macro
```
1. Do we really need to set "inc" and "lib"?
I would expect only  "incflag" and "libflag" would be needed

2. I think the naming of variables should follow existing conventions to make it easier to maintain.
For example: set MOD_LDFLAGS or ${MODULE}_LDFLAGS instead of "libflag"

yosefe commented on config/m4/module.m4:
```diff
@@ -281,27 +273,48 @@ AC_DEFUN([_CHECK_MODULE_PACKAGE],
 #                         [alt-dirs])
 #
+# Parameters:
+#   id           - module ID, used to generate runtime configuration parameters --with-id
+#   name         - readable module name
+#   prefix       - variable prefix, in success variables prefix_CFLAGS and
+#                  prefix_LIBS will store module specific flags
```
why id, name, and prefix can all be the same
(and prefix would be the capitalization of "id"/"name")?

yosefe commented on config/m4/module.m4:
```diff
@@ -281,27 +273,48 @@ AC_DEFUN([_CHECK_MODULE_PACKAGE],
+#                  prefix_LIBS will store module specific flags
+#   pkgname      - package name used to lookup module using pkg-config tool, optional
+#   help         - help string for module or empty argument to generate it automatically
+#   help-lib     - help string for module library dir or empty argument to generate
+#                  it automatically
+#   help-static  - help string for static module library or empty argument to
+#                  generate it automatically, if static link is not needed set it
+#                  to '-' value (see usage example)
```
maybe unite help, help-static and help-dynamic to one param?

yosefe commented on config/m4/module.m4:
```diff
@@ -281,27 +273,48 @@ AC_DEFUN([_CHECK_MODULE_PACKAGE],
+#   help-static  - help string for static module library or empty argument to
+#                  generate it automatically, if static link is not needed set it
+#                  to '-' value (see usage example)
+#   if-found     - macro called if module configuration found
```
action-if-found
action-if-notfound

yosefe commented on config/m4/module.m4:
```diff
@@ -281,27 +273,48 @@ AC_DEFUN([_CHECK_MODULE_PACKAGE],
+#                  to '-' value (see usage example)
+#   if-found     - macro called if module configuration found
+#   if-not-found - macro called if module configuration not found or user disabled module
+#   headers      - list of header files to check, coma-separated, optional
```
why need both headers and includes?

yosefe commented on config/m4/module.m4:
```diff
@@ -281,27 +273,48 @@ AC_DEFUN([_CHECK_MODULE_PACKAGE],
+#   headers      - list of header files to check, coma-separated, optional
+#   decls        - declarations to check, coma-separated, optional
+#   includes     - additional include directives to check headers and declarations, optional
+#   libs         - list of libraries to check, whitespace-separated
```
why some params are comma-separated and others are whitespace? its confusing IMO

yosefe commented on config/m4/module.m4:
```diff
@@ -281,27 +273,48 @@ AC_DEFUN([_CHECK_MODULE_PACKAGE],
+#                  check libraries, optional
+#   alt-dirs     - alternative directories to lookup: in case if user didn't
+#                  specify exact path to module then script will use this list of
+#                  directories, optional
```
pls document the side effects of this macros (which vars it defines)

Commit:
BUILD/CONFIG: added module check script

- updated docs
- minor script refactoring

gleon99 commented:
Hi @hoopoepg - we're cleaning up the UCX PR backlog and plan to close this PR in 7 days. Please let us know if you'd like to keep it open.
